boot/boot_test.go

```diff
@@ -3749,3 +3764,138 @@ func (s *bootKernelCommandLineSuite) TestCommandLineUpdateUC20TransitionFullExtr
+		// those are intentionally filled by the test
+		"snapd_extra_cmdline_args": "canary",
+		"snapd_full_cmdline_args":  "canary",
+	})
+	c.Assert(err, IsNil)
+	s.bootloader.SetBootVarsCalls = 0
+
+	restoreBootloaderNoPanic := s.bootloader.SetMockPanic("SetBootVars")
+
+	// transition to gadget with cmdline.extra
+	sf := snaptest.MakeTestSnapWithFiles(c, gadgetSnapYaml, [][]string{
+		{"cmdline.extra", "extra args"},
+	})
+
+	// let's panic on reseal first
+	resealPanic = true
+	c.Assert(func() {
+		boot.UpdateCommandLineForGadgetComponent(s.uc20dev, sf)
+	}, PanicMatches, "reseal panic")
+	c.Check(s.resealCalls, Equals, 1)
+	c.Check(s.resealCommandLines, DeepEquals, [][]string{{
+		// those come from boot chains which use predictable sorting
+		"snapd_recovery_mode=run static mocked panic=-1",
+		"snapd_recovery_mode=run static mocked panic=-1 extra args",
+	}})
+	// bootenv hasn't been updated yet
+	c.Check(s.bootloader.SetBootVarsCalls, Equals, 0)
+	// but modeenv has already been updated
+	m, err := boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(m.CurrentKernelCommandLines, DeepEquals, boot.BootCommandLines{
+		"snapd_recovery_mode=run static mocked panic=-1",
+		"snapd_recovery_mode=run static mocked panic=-1 extra args",
+	})
+
+	// REBOOT
+	resealPanic = false
+	err = boot.MarkBootSuccessful(s.uc20dev)
+	c.Assert(err, IsNil)
+	// we resealed after reboot, since modeenv was updated and carries the
+	// current command line only
+	c.Check(s.resealCalls, Equals, 2)
+	m, err = boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(m.CurrentKernelCommandLines, DeepEquals, boot.BootCommandLines{
+		"snapd_recovery_mode=run static mocked panic=-1",
+	})
+
+	// try the update again, but no panic in reseal this time
+	s.resealCalls = 0
+	s.resealCommandLines = nil
+	resealPanic = false
+	// but panic in set
+	c.Assert(func() {
+		boot.UpdateCommandLineForGadgetComponent(s.uc20dev, sf)
+	}, PanicMatches, "mocked reboot panic in SetBootVars")
+	c.Check(s.resealCalls, Equals, 1)
+	c.Check(s.resealCommandLines, DeepEquals, [][]string{{
+		// those come from boot chains which use predictable sorting
+		"snapd_recovery_mode=run static mocked panic=-1",
+		"snapd_recovery_mode=run static mocked panic=-1 extra args",
+	}})
+	// the call to bootloader wasn't counted, because it called panic
+	c.Check(s.bootloader.SetBootVarsCalls, Equals, 0)
+	m, err = boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(m.CurrentKernelCommandLines, DeepEquals, boot.BootCommandLines{
+		"snapd_recovery_mode=run static mocked panic=-1",
+		"snapd_recovery_mode=run static mocked panic=-1 extra args",
+	})
+
+	// REBOOT
+	err = boot.MarkBootSuccessful(s.uc20dev)
+	c.Assert(err, IsNil)
+	// we resealed after reboot again
+	c.Check(s.resealCalls, Equals, 2)
+	m, err = boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(m.CurrentKernelCommandLines, DeepEquals, boot.BootCommandLines{
+		"snapd_recovery_mode=run static mocked panic=-1",
+	})
+
```

Comment:
we should also check what happens if we reboot after setting the vars but before setting the task to done

Reply:
Updated. I've tried to add a devicestate level test, but could not come up with reasonable way to handle a panic in task runner. Instead I've added one more boot package test that panics after setting boot vars, which should mimic the scenario in which we reboot before marking the task as done.